src/main/java/org/example/bookstore/service/impl/BookServiceImpl.java

```diff
@@ -57,6 +77,34 @@ public void delete(Long id) {
     @Override
     public Page<BookDto> search(BookSearchParameters params, Pageable pageable) {
         Specification<Book> bookSpecification = bookSpecificationBuilder.build(params);
-        return bookRepository.findAll(bookSpecification, pageable).map(bookMapper::toDto);
+        Page<Book> bookPage = bookRepository.findAll(bookSpecification, pageable);
+        return bookPage.map(book -> {
+            BookDto dto = bookMapper.toDto(book);
+            setCategoryIds(dto, book);
+            return dto;
+        });
+    }
+
+    @Override
+    public Page<BookDtoWithoutCategoryIds> getByCategoriesId(Long id, Pageable pageable) {
+        return bookRepository.findAllByCategoriesId(id, pageable)
+            .map(bookMapper::toDtoWithoutCategories);
+    }
+
+    private void setCategoriesFromIds(Book book, CreateBookRequestDto bookDto) {
```

Comment:
you can use this colution

```suggestion
    private Set<Category> categoriesIdToCategories(List<Long> categories) {
        return categories.stream()
                .map(categoryRepository::getReferenceById)
                .collect(Collectors.toSet());
    }
```

and use it here
`book.setCategories(categoriesIdToCategories(bookDto.categoriesId()));`